libpromises/evalfunction.c

```diff
@@ -8451,6 +8451,230 @@ static FnCallResult FnCallNetworkConnections(EvalContext *ctx, ARG_UNUSED const
+        return NULL;
+    }
+
+    thread_data->result = FnReturnContext(success);
+    return NULL;
+}
+
+static FnCallResult FnCallIsReadable(ARG_UNUSED EvalContext *const ctx,
+                                     ARG_UNUSED const Policy *const policy,
+                                     const FnCall *const fp,
+                                     const Rlist *finalargs)
+{
+    assert(fp != NULL);
+    assert(fp->name != NULL);
+
+    struct IsReadableThreadData thread_data = {0};
+
+    if (finalargs == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function %s requires path as first argument",
+            fp->name);
+        return FnFailure();
+    }
+    thread_data.path = RlistScalarValue(finalargs);
+
+    long timeout = (finalargs->next == NULL) ? 3L
+                 : IntFromString(RlistScalarValue(finalargs->next));
+    assert(timeout >= 0);
+
+    if (timeout == 0) // Try read in main thread, possibly blocking forever
+    {
+        Log(LOG_LEVEL_DEBUG,
+            "Checking if file '%s' is readable in main thread, "
+            "possibly blocking forever.", thread_data.path);
+
+        char buf[1];
+        const int fd = open(thread_data.path, O_RDONLY);
+        if (fd < 0)
+        {
+            Log(LOG_LEVEL_DEBUG, "Failed to open file '%s': %s",
+                thread_data.path, GetErrorStr());
+            return FnReturnContext(false);
+        }
+        else if (read(fd, buf, sizeof(buf)) < 0)
+        {
+            Log(LOG_LEVEL_DEBUG, "Failed to read from file '%s': %s",
+                thread_data.path, GetErrorStr());
+            close(fd);
+            return FnReturnContext(false);
+        }
+
+        close(fd);
+        return FnReturnContext(true);
+    }
+    // Else try read in separate thread, possibly blocking for N seconds
+    Log(LOG_LEVEL_DEBUG,
+        "Checking if file '%s' is readable in separate thread, "
+        "possibly blocking for %ld seconds.", thread_data.path, timeout);
+
+    struct timespec timespec = {0};
+    xclock_gettime(CLOCK_REALTIME, &timespec);
+    timespec.tv_sec += timeout;
+
+    int ret = pthread_mutex_init(&(thread_data.mutex), NULL);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to initialize mutex: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    // Keep thread from doing its thing until we call
+    // pthread_cond_timedwait(3)
+    ret = pthread_mutex_lock(&(thread_data.mutex));
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to lock mutex: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_create(&(thread_data.thread), NULL,
+                         &IsReadableThreadRoutine, &thread_data);
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to create thread: %s",
+            GetErrorStrFromCode(ret));
+        return FnFailure();
+    }
+
+    ret = pthread_cond_init(&(thread_data.cond), NULL);
```

Comment:
Shouldn't this be done before the thread is started?

Reply:
Does not really matter. The thread is blocked on trying to acquire the mutex until after the condition is created. The thread uses the condition only after acquiring the mutex. Having `pthread_cond_init` and `pthread_cond_timedwait` consecutively makes the code more readable in my opinion.